muspinsim/celio.py

```diff
@@ -0,0 +1,382 @@
+           method
+
+        Returns the hamiltonian contributions defined by this system of spins and the
+        given interactions. In general these are split up per group of indices defined
+        in interactions to minimise the need of matrix exponentials.
+
+        Returns:
+            H_contribs {[CelioHContrib]} -- List of matrices representing contributions
+                                            to the total system hamiltonians referred to
+                                            in Celio's method as H_i
+        """
+
+        spin_indices = range(0, len(self._spinsys.spins))
+
+        H_contribs = []
+
+        for i in spin_indices:
+            # Only want to include each interaction once, will make the choice here to
+            # only add it to the H_i for the first particle listed in the interactions
+
+            # Find the terms that have the current spin as its first or only index
+            spin_ints = [term for term in (self._terms) if i == term.indices[0]]
+
+            # List of spin indices not included here
+            other_spins = list(range(0, len(self._spinsys.spins)))
+            other_spins.remove(i)
+
+            # Only include necessary terms
+            if len(spin_ints) != 0:
+                # Sum matrices with the same indices so we avoid lots of matrix
+                # exponentials
+                for indices, group in itertools.groupby(
+                    spin_ints, lambda term: term.indices
+                ):
+
+                    grouped_spin_ints = list(group)
+                    H_contrib = np.sum([term.matrix for term in grouped_spin_ints])
+
+                    # Find indices of spins not involved in the current interactions
+                    other_spins_copy = other_spins.copy()
+                    for term in grouped_spin_ints:
+                        for j in term.indices:
+                            if j in other_spins_copy:
+                                other_spins_copy.remove(j)
+
+                    other_dimension = np.product(
+                        [self._spinsys.dimension[j] for j in other_spins_copy]
+                    )
+
+                    # Detect Quadrupolar terms which will have the same index twice
+                    indices = list(indices)
+                    if len(indices) == 2 and indices[0] == indices[1]:
+                        # Only include one of them for the ordering
+                        indices.pop()
+
+                    # Order in which kronecker products will be performed in Celio's
+                    # method
+                    spin_order = list(indices) + other_spins_copy
+
+                    # Order we need to permute in order to obtain the same order as was
+                    # given in the input
+                    permute_order = np.zeros(len(spin_order), dtype=np.int32)
+                    for i, value in enumerate(spin_order):
+                        permute_order[value] = i
```

Comment:
I _think_ what you're doing here could be achieved by `permute_order = np.argsort(spin_order)`?
https://numpy.org/doc/stable/reference/generated/numpy.argsort.html

Reply:
Great spot, its very slightly faster too.